### ISSUE TITLE
Generic function has function in props that does not detect return type when param added

### ISSUE THREAD

**Issue by 700software** [microsoft/TypeScript#48279]
# Bug Report

### 🔎 Search Terms

Posted question to StackOverflow:
[Generic React/Typescript function has function in props but does not detect return type when param added](https://stackoverflow.com/questions/71401721/generic-react-typescript-function-has-function-in-props-but-does-not-detect-retu)
but it occurs to me this may be a bug!?

### 🕗 Version & Regression Information

- Tested in TypeScript 4.5.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAZxgWwA4BsbBgUwBMAeAFQD4AKDAJzg2QC5EBvRAc3zHxoEMo4NZpQCUiALzlEpADSIa+CPhgA3HsKjNSYyYl5gAnogC+YlgFgAUMatXQkWAkQB3ABb8AkgHV9UMlVp6JlYOLh5+QWEAIzgo5n0DHSlZeUVlNSFESk1pJL1DEzMrG0s7cGh4JDAEAHkogCtFPwpKTm4+AUzKGLj8xIlkuQUlVXUsnO0BvqLrW0tUTBw8Iko2NvDO4TyARgAmAGYhtNGuqDyIBGQ4LHwAOiw4dmzEAGpEXbFTKzdPHzAoVahdoRLo9PI9RAAfkQ20QzF2RxGGQ050u1zuDyeUFe70+Iis1TAdUa0Eo3Vi4NiUJhcPecmyqLAVxu90ezzeH3xliAA)

### 💻 Code

```ts
function simplified<T>(props: { generator: () => T, receiver: (t: T) => any }) {
}

function whatIWant<T>(props: { generator: (bob: any) => T, receiver: (t: T) => any }) {
}

function nonObject<T>(generator: (bob: any) => T, receiver: (t: T) => any) {
}

simplified({ generator: () => 123, receiver: (t) => console.log(t + 2) })
whatIWant({ generator: (bob) => bob ? 1 : 2, receiver: (t) => console.log(t + 2) })
nonObject((bob) => bob ? 1 : 2, (t) => console.log(t + 2))
```

### 🙁 Actual behavior

The `whatIWant` function is showing an error because it thinks `t` is `unknown` type.

The other two (`simplified` and `nonObject` are correctly seeing `t` as a `number` type.

![image](https://user-images.githubusercontent.com/20981023/158486845-c96dcb0c-8c82-47f9-a272-04c437ac39ed.png)

### 🙂 Expected behavior

Above code (all three cases) should have no errors.

I'm particularly confused by the difference between `simplified` and `whatIWant`. Why is this correct behaviour, or is it a bug?

**Comment by RyanCavanaugh** [comment]
See #47599
